### DISCUSSION TITLE
Unexpected Package Name Behavior

### DISCUSSION THREAD

**Discussion by lbettels** [soot-oss/SootUp#812] [Q&A] [answered]
When i load classes into a view, it seems like SootUp omits the longest common prefix in the package name.
So if i have two classes A and B, with A being in the package a.b and B being in the package a.b.c, SootUp drops the prefix a.b.
So when i load the classes, A shows up in package "" and B shows up in package "c".

This is problematic  when i try to build a ``` JimpleBasedInterproceduralCFG``` .
If A uses function foo of B, then the MethodSignature will show up as "a.b.c.B foo()", but the view only knows "c.B foo()".
The algorithm therefore fails to include the necessary functions because it tries to load functions with the prefix a.b.c and not the prefix c.

**Comment by lbettels** [comment]
1. I use SootUp 1.1.2 compiled from source from branch 71d550872 Merge pull request #791 from soot-oss/FieldRefValidator_changes

The directory structure in the out directory used in the next code snippet is pretty straight forward
<pre>
     out
     |-> PackageTest.class
     |-> SubPackage
              |-> SubPackageTest.class
</pre>

```java
public class PackageNameTest {
    public static void main(String[] args) {
        Path pathToSource = Paths.get("src/main/resources/sample_files/out/");
        JavaLanguage language = new JavaLanguage(21);
        List<AnalysisInputLocation<? extends JavaSootClass>> inputLocations = new ArrayList<>();
        inputLocations.add(new JavaClassPathAnalysisInputLocation(pathToSource.toString()));
        JavaView view = new JavaView(inputLocations);
        List<?> classes = (List<?>) view.getClasses();
    }
}
```
This is what the source files of the two classes look like:
1. PackageTest.java:
```java
package packageTest;
import packageTest.SubPackage.SubPackageTest;
public class PackageTest {
    public PackageTest(){}
    public void myFunction(){
        SubPackageTest sub = new SubPackageTest();
        System.out.println(sub.sub());
    }
}
```
2. SubPackageTest.java
```java
package packageTest.SubPackage;

public class SubPackageTest {
    public SubPackageTest(){}
    public String sub(){
        return "sub";
    }
}
```
The two classes successfully compile together and keep their package declaration in the class files.
Screenshots:
1. PackageTest.java dropping its package entirely
![packagetest](https://github.com/soot-oss/SootUp/assets/43786637/1e081299-6910-4138-84c4-7e5cc924ddfd)
2. SubPackageTest.java dropping its prefix
![subpackagetest](https://github.com/soot-oss/SootUp/assets/43786637/e4ed5425-f128-4322-b813-57c139985488)

Thank you for answering to my issue so quickly!

**Reply by JonasKlauke** [reply, marked as answer]
The packages are extracted by the root folder. In your example the root is out. Therefore, PackageTest has no package because it is directly in the root folder.
Set the root folder in the Inputlocation to the start of the package names.
You are missing the packageTest folder that contains everything. if the correct fully qualified name is
packageTest.Subpackage.SubPackageTest then your folder structure should look like this.

out     
|-> packageTest
     |-> PackageTest.class
     |-> SubPackage
              |-> SubPackageTest.class

SootUp is not consistent because the bytecode uses the correct fully qualified name because it is contained in the bytecode. Therefore, your inputlocation needs to point to a jar or a folder that contains the entire internal structure, since the fully qualified name is decided by the path to a class file, since the inputlocation works like a classloader
